homeassistant/components/enocean/cover.py

```diff
@@ -0,0 +1,267 @@
+
+        telegram = [0xD2, 100, 0, 0, 1]
+        telegram.extend(self._sender_id)
+        telegram.extend([0x00])
+        self.send_telegram(telegram)
+
+    def set_cover_position(self, **kwargs) -> None:
+        """Set the cover position."""
+        self._state_changed_by_command = True
+
+        if kwargs[ATTR_POSITION] == self._position:
+            self._is_opening = False
+            self._is_closing = False
+        elif kwargs[ATTR_POSITION] > self._position:
+            self._is_opening = True
+            self._is_closing = False
+        elif kwargs[ATTR_POSITION] < self._position:
+            self._is_opening = False
+            self._is_closing = True
+
+        self.start_or_feed_watchdog()
+
+        telegram = [0xD2, 100 - kwargs[ATTR_POSITION], 0, 0, 1]
+        telegram.extend(self._sender_id)
+        telegram.extend([0x00])
+        self.send_telegram(telegram)
+
+    def stop_cover(self, **kwargs) -> None:
+        """Stop any cover movement."""
+        self.stop_watchdog()
+        self._state_changed_by_command = True
+        self._is_opening = False
+        self._is_closing = False
+
+        telegram = [0xD2, 2]
+        telegram.extend(self._sender_id)
+        telegram.extend([0x00])
+        self.send_telegram(telegram)
+
+    def value_changed(self, packet):
+        """Fire an event with the data that have changed.
+
+        This method is called when there is an incoming packet associated
+        with this platform.
+        """
+        # position is inversed in Home Assistant and in EnOcean:
+        # 0 means 'closed' in Home Assistant and 'open' in EnOcean
+        # 100 means 'open' in Home Assistant and 'closed' in EnOcean
+
+        new_position = 100 - packet.data[1]
+
+        if self._position is not None:
+            if self._state_changed_by_command:
+                self._state_changed_by_command = False
+            elif new_position in (0, 100, self._position):
+                self._is_opening = False
+                self._is_closing = False
+                self.stop_watchdog()
+            elif new_position > self._position:
+                self._is_opening = True
+                self._is_closing = False
+                self.start_or_feed_watchdog()
+            elif new_position < self._position:
+                self._is_opening = False
+                self._is_closing = True
+                self.start_or_feed_watchdog()
+
+        self._position = new_position
+        if self._position == 0:
+            self._is_closed = True
+        else:
+            self._is_closed = False
+
+        self.schedule_update_ha_state()
+
+    def send_telegram(self, data):
+        """Send a telegram via the EnOcean dongle to only this device."""
+        # optional data contains: number of subtelegrams (fixed to 3 for sending),
+        # destination id, max dBm (0xFF) for sending and security level 0
+        packet = Packet(0x01, data=data, optional=[3] + self.dev_id + [0xFF, 0])
```

Comment:
Here there are also protocol details.

Reply:
Agreed. Fortunately the used EnOcean-library already has better high-level functions, please have a look at commit [3dd3a08](https://github.com/home-assistant/core/pull/75356/commits/3dd3a085cc5c62c775f5783881ba0faccb6c4836).